DiscriminatedOnions/Result.cs

```diff
@@ -73,6 +73,12 @@ public static class Result
     public static Result<U, TError> Bind<T, TError, U>(this Result<T, TError> result, Func<T, Result<U, TError>> binder) =>
         result.Match(Error<U, TError>, binder);
 
+    public static async Task<Result<U, TError>> BindAsync<T, TError, U>(this Task<Result<T, TError>> result, Func<T, Task<Result<U, TError>>> binder)
```

Comment:
I'm not sure we need a separate overload for this: see the comment in the related test.
We already have the four overload of Pipe that can be used to resolve the impedance mismatch between functions accepting or returning tasks that must be awaited.
The idea is that, unless strictly required, you don't need to bother to write every possible combination of functions being Task'y in either its input or output.
I may miss something, but what is the use case that raised your need?